client/my-sites/controller.js

```diff
@@ -324,38 +334,40 @@ export function siteSelection( context, next ) {
 		return next();
 	}
 
-	/**
-	 * If the user has only one site, redirect to the single site
-	 * context instead of rendering the all-site views.
+	/*
+	 * If the user has only one site, redirect to the single site context instead of
+	 * rendering the all-site views.
 	 *
-	 * If the /me/sites API endpoint hasn't returned yet, postpone the redirect until
-	 * after the sites data are available by scheduling a `SITES_ONCE_CHANGED` callback.
+	 * If the primary site is not yet available in Redux state, initiate a fetch and postpone the
+	 * redirect until the fetch is complete. (while the primary site ID is a property of the
+	 * current user object and therefore always available, we need to fetch the site info in order
+	 * to convert the site ID to the site slug that will be part of the redirect URL)
 	 */
 	if ( hasOneSite && ! siteFragment ) {
-		const redirectToPrimary = () => {
-			const primarySiteSlug = getPrimarySiteSlug( getState() );
-			let redirectPath = `${ context.pathname }/${ primarySiteSlug }`;
-
-			redirectPath = context.querystring
-				? `${ redirectPath }?${ context.querystring }`
-				: redirectPath;
+		const primarySiteId = getPrimarySiteId( getState() );
 
+		const redirectToPrimary = primarySiteSlug => {
+			let redirectPath = `${ context.pathname }/${ primarySiteSlug }`;
+			if ( context.querystring ) {
+				redirectPath += `?${ context.querystring }`;
+			}
 			page.redirect( redirectPath );
 		};
 
-		const hasInitialized = getSites( getState() ).length;
-		if ( hasInitialized ) {
-			if ( getPrimarySiteSlug( getState() ) ) {
-				redirectToPrimary();
-			} else {
-				// If the primary site does not exist, skip redirect
-				// and display a useful error notification
-				showMissingPrimaryError( currentUser, dispatch );
-			}
+		const primarySiteSlug = getSiteSlug( getState(), primarySiteId );
+		if ( primarySiteSlug ) {
+			redirectToPrimary( primarySiteSlug );
 		} else {
-			dispatch( {
-				type: SITES_ONCE_CHANGED,
-				listener: redirectToPrimary,
+			// Fetch the primary site by ID and then try to determine its slug again.
+			dispatch( requestSite( primarySiteId ) ).then( () => {
```

Comment:
what should we do if the promise rejects?

Reply:
The one returned from `requestSite` never does. `requestSite` [catches all errors and returns a promise that always resolves](https://github.com/Automattic/wp-calypso/blob/cd37e5898296583ccc1256d0f1f3228b50e41bf5/client/state/sites/actions.js#L83-L104). Very similar to a `finally` block.

Note that the `then` handler handles both success (site is there) and error (not there) conditions. If `requestSite` was changed to return a a rejected promise, these would be two independed handlers.

Reply:
👍